pkg/issuer/acme/dns/azuredns/azuredns.go

```diff
@@ -73,140 +73,64 @@ func NewDNSProviderCredentials(environment, clientID, clientSecret, subscription
-	// NOTE: all related environment variables are described here: https://azure.github.io/azure-workload-identity/docs/installation/mutating-admission-webhook.html
-	oauthConfig, err := adal.NewOAuthConfig(env.ActiveDirectoryEndpoint, os.Getenv("AZURE_TENANT_ID"))
-	if err != nil {
-		return nil, fmt.Errorf("failed to retrieve OAuth config: %v", err)
-	}
-
-	jwt, err := os.ReadFile(os.Getenv("AZURE_FEDERATED_TOKEN_FILE"))
-	if err != nil {
-		return nil, fmt.Errorf("failed to read a file with a federated token: %v", err)
-	}
-
-	// AZURE_CLIENT_ID will be empty in case azure.workload.identity/client-id annotation is not set
-	// Also, some users might want to use a different MSI for a particular DNS zone
-	// Thus, it's important to offer optional ClientID overrides
-	clientID := os.Getenv("AZURE_CLIENT_ID")
-	if options.ClientID != "" {
-		clientID = options.ClientID
-	}
-
-	token, err := adal.NewServicePrincipalTokenFromFederatedToken(*oauthConfig, clientID, string(jwt), env.ResourceManagerEndpoint)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create a workload identity token: %v", err)
-	}
-
-	return token, nil
-}
-
-func getAuthorization(env azure.Environment, clientID, clientSecret, subscriptionID, tenantID string, ambient bool, managedIdentity *cmacme.AzureManagedIdentity) (*adal.ServicePrincipalToken, error) {
+func getAuthorization(clientOpt policy.ClientOptions, clientID, clientSecret, tenantID string, ambient bool, managedIdentity *cmacme.AzureManagedIdentity) (azcore.TokenCredential, error) {
 	if clientID != "" {
 		logf.Log.V(logf.InfoLevel).Info("azuredns authenticating with clientID and secret key")
-		oauthConfig, err := adal.NewOAuthConfig(env.ActiveDirectoryEndpoint, tenantID)
-		if err != nil {
-			return nil, err
-		}
-		spt, err := adal.NewServicePrincipalToken(*oauthConfig, clientID, clientSecret, env.ResourceManagerEndpoint)
+		cred, err := azidentity.NewClientSecretCredential(tenantID, clientID, clientSecret, &azidentity.ClientSecretCredentialOptions{ClientOptions: clientOpt})
 		if err != nil {
 			return nil, err
 		}
-		return spt, nil
+		return cred, nil
 	}
 	logf.Log.V(logf.InfoLevel).Info("No ClientID found: attempting to authenticate with ambient credentials (Azure Workload Identity or Azure Managed Service Identity, in that order)")
 	if !ambient {
 		return nil, fmt.Errorf("ClientID is not set but neither `--cluster-issuer-ambient-credentials` nor `--issuer-ambient-credentials` are set. These are necessary to enable Azure Managed Identities")
 	}
 
-	opt := adal.ManagedIdentityOptions{}
-
-	if managedIdentity != nil {
-		opt.ClientID = managedIdentity.ClientID
-		opt.IdentityResourceID = managedIdentity.ResourceID
-	}
-
 	// Use Workload Identity if present
 	if os.Getenv("AZURE_FEDERATED_TOKEN_FILE") != "" {
-		spt, err := getFederatedSPT(env, opt)
-		if err != nil {
-			return nil, err
+		wcOpt := &azidentity.WorkloadIdentityCredentialOptions{
+			DisableInstanceDiscovery: true,
```

Comment:
Not sure whether this option should be `true` or `false`.

Reply:
@phillebaba or @weisdd Can either of answer this?

 * https://learn.microsoft.com/en-us/dotnet/api/azure.identity.environmentcredentialoptions.disableinstancediscovery?view=azure-dotnet

> Gets or sets the setting which determines whether or not instance discovery is performed when attempting to authenticate. Setting this to true will completely disable both instance discovery and authority validation. This functionality is intended for use in scenarios where the metadata endpoint cannot be reached, such as in private clouds or Azure Stack. The process of instance discovery entails retrieving authority metadata from https://login.microsoft.com/ to validate the authority. By setting this to true, the validation of the authority is disabled. As a result, it is crucial to ensure that the configured authority host is valid and trustworthy."